source/material_model/prescribed_viscosity.cc

```diff
@@ -0,0 +1,206 @@
+/*
+  Copyright (C) 2011 - 2023 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/material_model/prescribed_viscosity.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <array>
+
+#include <utility>
+#include <limits>
+
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    template <int dim>
+    void
+    PrescribedViscosity<dim>::initialize()
+    {
+      base_model->initialize();
+    }
+
+
+
+    template <int dim>
+    void
+    PrescribedViscosity<dim>::update()
+    {
+      base_model->update();
+
+      // we get time passed as seconds (always) but may want
+      // to reinterpret it in years
+      if (this->convert_output_to_years())
+        {
+          prescribed_viscosity_indicator_function.set_time (this->get_time() / year_in_seconds);
+          prescribed_viscosity_function.set_time(this->get_time() / year_in_seconds);
+        }
+      else
+        {
+          prescribed_viscosity_indicator_function.set_time (this->get_time());
+          prescribed_viscosity_function.set_time (this->get_time());
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    PrescribedViscosity<dim>::evaluate(const typename Interface<dim>::MaterialModelInputs &in,
+                                       typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      base_model->evaluate(in,out);
+      for (unsigned int i=0; i<out.n_evaluation_points(); ++i)
+        {
+          if (in.requests_property(MaterialProperties::viscosity))
+            {
+              const double indicator = prescribed_viscosity_indicator_function.value(in.position[i]);
+              const double viscosity = prescribed_viscosity_function.value(in.position[i]);
+
+              if (indicator > 0.5)
+                {
+                  out.viscosities[i] = viscosity;
+                }
+            }
```

Comment:
Maybe like this (saves a tiny amount of time):
```suggestion
          if (in.requests_property(MaterialProperties::viscosity))
            {
              const double indicator = prescribed_viscosity_indicator_function.value(in.position[i]);

              if (indicator > 0.5)
                {
                  out.viscosities[i] = prescribed_viscosity_function.value(in.position[i]);
                }
            }
```